File: src/com/sam3125c/algorithm/Sort.java
/**
 * Yang Rong Individual Service
 * All Rights Reserved @2020
 */
package com.sam3125c.algorithm;

/**
 * 排序算法工具类。
 *
 * @author Yang Rong
 * @version 1.0
 * @email samyang3125c@gmail.com
 * @create 2020-02-26
 */
public class Sort {

    /**
     * 冒泡排序。
     *
     * @param array
     * @return int[]
     * @author Yang Rong
     * @create 2020/2/27
     */
    public static int[] bubbleSort(int[] array) {
        int len = array.length;
        for (int i = 0; i < len - 1; i++) {
            for (int j = 0; j < len - 1 - i; j++) {
                if (array[j] > array[j + 1]) {
                    int tmp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = tmp;
                }
            }
        }
        return array;
    }

    /**
     * 插入排序。
     *
     * @param array
     * @return int[]
     * @author Yang Rong
     * @create 2020/2/27
     */
    public static int[] insertSort(int[] array) {
        int len = array.length;
        for (int i = 1; i < len; i++) {
            int insert = array[i];
            int j = i - 1;
            for (; j >= 0; j--) {
                if (array[j] > insert) {
                    array[j + 1] = array[j];
                } else {
                    break;
                }
            }
            array[j + 1] = insert;
        }
        return array;
    }

    /**
     * 选择排序。
     *
     * @param array
     * @return int[]
     * @author Yang Rong
     * @create 2020/2/27
     */
    public static int[] choiceSort(int[] array) {
        int len = array.length;
        for (int i = 0; i < len; i++) {
            int min = i;
            for (int j = i; j < len; j++) {
                if (array[j] < array[min]) {
                    min = j;
                }
            }
            int temp = array[i];
            array[i] = array[min];
            array[min] = temp;
        }
        return array;
    }

    /**
     * 归并排序。
     *
     * @param array
     * @return int[]
     * @author Yang Rong
     * @create 2020/3/3
     */
    public static int[] mergeSort(int[] array) {
        recursiveInMergeSort(array, 0, array.length - 1);
        return array;
    }

    /**
     * 归并排序的递归部分。
     *
     * @param array
     * @param p
     * @param r
     * @return void
     * @author Yang Rong
     * @create 2020/3/3
     */
    private static void recursiveInMergeSort(int[] array, int p, int r) {
        // 无法进一步分解了。
        if (p >= r) return;
        // 分解逻辑，取中值。
        int q = p + (r - p) / 2;
        // 递归分解中值左半部分。
        recursiveInMergeSort(array, p, q);
        // 递归分解中值右半部分。
        recursiveInMergeSort(array, q + 1, r);
        // 有序合并左右部分。
        mergeInSortMerge(array, p, q, r);
    }

    /**
     * 归并排序的合并部分。
     *
     * @param array
     * @param p
     * @param q
     * @param r
     * @return void
     * @author Yang Rong
     * @create 2020/3/3
     */
    private static void mergeInSortMerge(int[] array, int p, int q, int r) {
        // 左半部分起始。
        int i = p;
        // 右半部分起始。
        int j = q + 1;
        // 临时数组相关。
        int k = 0;
        int[] tmp = new int[r - p + 1];

        // 左右半部的比较，从小到大放入临时数组。直到一个半部放完。
        while (i <= q && j <= r) {
            if (array[i] <= array[j]) {
                tmp[k++] = array[i++];
            } else {
                tmp[k++] = array[j++];
            }
        }

        // 假设 while 被 j > r 破坏，则未放完的是 i,q。
        // 如果并不是，则 j <= r，那么未放完的是 j,r。
        int start = i;
        int end = q;
        if (j <= r) {
            start = j;
            end = r;
        }
        // 确定后，继续放未放完的半部。
        while (start <= end) {
            tmp[k++] = array[start++];
        }

        // 将临时数组中数据，拷贝到原数组。
        for (int c = 0; c < r - p + 1; c++) {
            array[p + c] = tmp[c];
        }
    }

    /**
     * 快速排序。
     *
     * @param array
     * @return int[]
     * @author Yang Rong
     * @create 2020/3/3
     */
    public static int[] quickSort(int[] array) {
        recursiveInQuickSort(array, 0, array.length - 1);
        return array;
    }

    /**
     * 快速排序的递归部分。
     *
     * @param array
     * @param p
     * @param r
     * @return void
     * @author Yang Rong
     * @create 2020/3/3
     */
    private static void recursiveInQuickSort(int[] array, int p, int r) {
        // 无法进一步分区了。
        if (p >= r) return;
        // 获取逻辑，取得分区的下标。
        int q = partitionInQuickSort(array, p, r);
        // 递归分区小于分区下标的部分。
        recursiveInQuickSort(array, p, q - 1);
        // 递归分区大于分区下标的部分。
        recursiveInQuickSort(array, q + 1, r);
    }

    /**
     * 快速排序的分区部分。
     *
     * @param array
     * @param p
     * @param r
     * @return int
     * @author Yang Rong
     * @create 2020/3/3
     */
    private static int partitionInQuickSort(int[] array, int p, int r) {
        // 一般取 p,r 的最后一个元素为 pivot。
        int pivot = array[r];

        // 指针 i 用于维护大小区间的分界点，并且挂在大于 pivot 的第一个元素上。
        int i = p;
        // 指针 j 用于遍历。
        for (int j = p; j < r - 1; j++) {
            // 如果遍历的元素小于 pivot，则将该元素和 i 上元素互换。
            // 当 i 自增后，i 元素左边的元素都是小于 pivot 的。
            // 并且 i 继续保持挂在大于 pivot 的第一个元素上。
            if (array[j] < pivot) {
                int tmp = array[i];
                array[i] = array[j];
                array[j] = tmp;
                i++;
            }
        }

        // 最后让 i 与 pivot 互换，让 pivot 去到真正分割大小区间的位置。
        array[r] = array[i];
        array[i] = pivot;

        //此时下标 i 即为 pivot 下标。
        return i;
    }
}

File: python/algorithm/sort.py
"""
Yang Rong Individual Service
All Rights Reserved @2021

排序相关算法的实现。

@author: Yang Rong
@version 1.0
@email: samyang3125c@gmail.com
@create: 2021-04-05
"""


def bubble_sort(arr):
    """
    冒泡排序的实现。
    :param arr: list to sort.
    :return: list: sorted arr.
    """
    _len = len(arr)
    for i in range(_len - 1):
        for j in range(_len - 1 - i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr


def insert_sort(arr):
    """
    插入排序的实现。
    :param arr: list to sort.
    :return: list: sorted arr.
    """
    _len = len(arr)
    for i in range(1, _len):
        insert = arr[i]
        pos = i - 1
        while pos >= 0 and arr[pos] > insert:
            arr[pos + 1] = arr[pos]
            pos -= 1
        arr[pos + 1] = insert
    return arr


def select_sort(arr):
    """
    选择排序的实现。
    :param arr: list to sort.
    :return: list: sorted arr.
    """
    _len = len(arr)
    for i in range(_len):
        pos = i
        for j in range(i, _len):
            if arr[j] < arr[pos]:
                pos = j
        arr[i], arr[pos] = arr[pos], arr[i]
    return arr


def merge_sort(arr):
    """
    归并排序的实现。
    :param arr: list to sort.
    :return: list: sorted arr.
    """

    def partition(array, left, right):
        if left >= right:
            return
        mid = left + ((right - left) >> 1)
        # 递归分区。
        partition(array, left, mid)
        partition(array, mid + 1, right)
        # 分到无法拆分之后，分区天然有序，然后有序合并各分区。
        merge(array, left, mid, right)

    def merge(array, left, mid, right):
        # 直到某个分区所有的元素均完成排序。
        i = left
        j = mid + 1
        tmp = []
        while i <= mid and j <= right:
            if array[i] < array[j]:
                tmp.append(array[i])
                i += 1
            else:
                tmp.append(array[j])
                j += 1
        # 将未完成排序的那个分区剩余的元素进行填充。
        if i > mid:
            start, end = j, right
        else:
            start, end = i, mid
        for E in array[start: end + 1]:
            tmp.append(E)
        # 将排序完的数据按顺序拷回原数组。
        for offset in range(right - left + 1):
            array[left + offset] = tmp[offset]

    partition(arr, 0, len(arr) - 1)
    return arr


def quick_sort(arr):
    """
    快速排序的实现。
    :param arr: list to sort.
    :return: list: sorted arr.
    """

    def partition(array, left, right):
        if left >= right:
            return
        p = quick(array, left, right)
        partition(array, left, p - 1)
        partition(array, p + 1, right)

    def quick(array, left, right):
        # 选取分界元素和分界游标。
        pivot = array[right]
        p = left
        # 与分界元素比较，分界游标左边均小于它、右边和自身均大于它。
        for i in range(left, right):
            if array[i] < pivot:
                array[p], array[i] = array[i], array[p]
                p += 1
        # 将分界元素和分界游标元素对调。
        array[p], array[right] = pivot, array[p]
        return p

    partition(arr, 0, len(arr) - 1)
    return arr


def heap_sort(arr):
    """
    堆排序的实现。
    :param arr: list to sort.
    :return: list: sorted arr.
    """

    def heapify(data, index, size=None):
        if not size:
            size = len(data)
        while True:
            # bigger_pos 引入可以帮助左右子节点的比较，还可以用作终止判断。
            bigger_pos = index
            if 2 * index < size and data[index] < data[2 * index]:
                bigger_pos = 2 * index
            if 2 * index + 1 < size and data[bigger_pos] < data[2 * index + 1]:
                bigger_pos = 2 * index + 1
            if bigger_pos == index:
                break
            data[index], data[bigger_pos] = data[bigger_pos], data[index]
            index = bigger_pos

    def build(data):
        data.insert(0, None)
        # 满二叉树的第 n/2 + 1 至 n 个节点是叶子结点，循环从 n/2 个开始。
        # 加入哨兵节点后 len(data) - 1，因此是 int((len(data) - 1) / 2)。
        # 第几个节点比数组的下标多一，而哨兵节点又需要补一。int((len(data) - 1) / 2) - 1 + 1。
        for i in range(int((len(data) - 1) / 2), 0, -1):
            heapify(data, i)

    # 待排序数组的堆化处理。
    build(arr)
    # 不断处理大顶堆堆顶，原地排序。
    for j in range(len(arr) - 1, 0, -1):
        arr[1], arr[j] = arr[j], arr[1]
        heapify(arr, 1, j)
    return arr[1:]


if __name__ == '__main__':
    sorted_arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    print(f'bubble_sort() is {"ok" if bubble_sort([4, 3, 5, 7, 8, 1, 2, 9, 6]) == sorted_arr else "not ok"}.')
    print(f'insert_sort() is {"ok" if insert_sort([4, 3, 5, 7, 8, 1, 2, 9, 6]) == sorted_arr else "not ok"}.')
    print(f'select_sort() is {"ok" if select_sort([4, 3, 5, 7, 8, 1, 2, 9, 6]) == sorted_arr else "not ok"}.')
    print(f'merge_sort() is {"ok" if merge_sort([4, 3, 5, 7, 8, 1, 2, 9, 6]) == sorted_arr else "not ok"}.')
    print(f'quick_sort() is {"ok" if quick_sort([4, 3, 5, 7, 8, 1, 2, 9, 6]) == sorted_arr else "not ok"}.')
    print(f'heap_sort() is {"ok" if heap_sort([4, 3, 5, 7, 8, 1, 2, 9, 6]) == sorted_arr else "not ok"}.')
